pkg/kafka/auth/scram.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	"hash"
+
+	"github.com/Shopify/sarama"
+	"github.com/pkg/errors"
+	SCRAM "github.com/xdg/scram"
+)
+
+// scramConfig describes the configuration properties required for the SCRAM handshake
+// between the collector and the pipeline
+type scramConfig struct {
+	userName  string `mapstructure:"userName"`
+	password  string `mapstructure:"password"`
+	algorithm string `mapstructure:"algorithm"`
+}
+
+func setSCRAMConfiguration(config *scramConfig, saramaConfig *sarama.Config) error {
+	saramaConfig.Net.SASL.Enable = true
+	saramaConfig.Net.SASL.User = config.userName
+	saramaConfig.Net.SASL.Password = config.password
+
+	if config.algorithm == "sha512" {
+		saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &SCRAMClient{HashGeneratorFcn: SHA512} }
+		saramaConfig.Net.SASL.Mechanism = sarama.SASLMechanism(sarama.SASLTypeSCRAMSHA512)
+	} else if config.algorithm == "sha256" {
+		saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &SCRAMClient{HashGeneratorFcn: SHA256} }
+		saramaConfig.Net.SASL.Mechanism = sarama.SASLMechanism(sarama.SASLTypeSCRAMSHA256)
+	} else {
+		return errors.Errorf("invalid SHA algorithm '%s': can be either 'sha256' or 'sha512'", config.algorithm)
+	}
+
+	return nil
+}
+
+// SCRAMClient takes a dereference to a scram client/conversation and a hash function
+type SCRAMClient struct {
+	*SCRAM.Client
+	*SCRAM.ClientConversation
+	SCRAM.HashGeneratorFcn
```

Comment:
Is there any need for embedding these? If not, it's better to give them private field names, because otherwise all their methods are leaking into the `scramClient` API.

Reply:
Afaik they're based in the interface for the sarama config to begin using Scram and need to be embedded for the Begin, Step and Done. Let me check and I'll reply to this comment.